src/solve/fulfill.rs

```diff
@@ -312,24 +307,29 @@ impl<'s> Fulfill<'s> {
                             }
                         }
 
-                        solution.is_ambig()
+                        (solution.is_ambig(), solution.cannot_be_proven())
                     }
                     Obligation::Refute(ref goal) => {
-                        self.refute(goal)? == NegativeSolution::Ambiguous
+                        let answer = self.refute(goal)?;
+                        (answer == NegativeSolution::Ambiguous, answer == NegativeSolution::CannotProve)
                     }
                 };
 
                 if ambiguous {
                     debug!("ambiguous result: {:?}", obligation);
                     obligations.push(obligation);
                 }
+
+                // If one of the obligations cannot be proven, then the whole goal
+                // cannot be proven either.
```

Comment:
Yeah exactly. I'm adding a bit of documentation here too. Notice that for the same reason, in the case where the unification set `cannot_prove = true`, we don't return `CannotProve` early here:
[https://github.com/scalexm/chalk/blob/cannot-prove/src/solve/solver.rs#L214](https://github.com/scalexm/chalk/blob/cannot-prove/src/solve/solver.rs#L214)